reps/2023-06-06-simplify-sync.md

```diff
@@ -0,0 +1,109 @@
+
+Both of these have been significant pain points over the past year, based both on user and internal feedback. This shouldn't be too surprising, as the persistent storage APIs have never been significantly revised before, beyond introduction of the new `storage_path` option, and similarly the Checkpoint API is in beta.
+
+### Should this change be within `ray` or outside?
+
+main `ray` project. Changes are made to Ray Train and Tune.
+
+## Stewardship
+
+### Required Reviewers
+The proposal will be open to the public, but please suggest a few experienced Ray contributors in this technical domain whose comments will help this proposal. Ideally, the list should include Ray committers.
+
+@matthewdeng, @krfricke
+
+### Shepherd of the Proposal (should be a senior committer)
+To make the review process more productive, the owner of each proposal should identify a **shepherd** (should be a senior Ray committer). The shepherd is responsible for working with the owner and making sure the proposal is in good shape (with necessary information) before marking it as ready for broader review.
+
+@pcmoritz
+
+## User Story
+
+Persistence will center around the recently introduced `storage_path` option.
+
+### Persistent trial directory
+
+Ray Train provides the abstraction of a persistent trial directory. Each trial has its current working directory set to a trial specific directory (e.g., `~/ray_results/experiment_1/trial_1`). Train will synchronize this trial directory with persistent storage specified via the `storage_path` option.
+
+This directory is persistent since Train is syncing it to persistent storage for the user. The experiment driver will record experiment results and other metadata to a similar directory on the head node, which will also synchronize to persistent storage.
+
+![overview](user-story.png)
+
+### Checkpoints
+
+While trials can write arbitrary data to the persistent trial directory, they cannot be guaranteed they can read it back on checkpoint/restore. Data that is needed for resume must be recorded via the `train.report(checkpoint)` API.
+
+For example, this is how you could record a Torch checkpoint:
+
+```python
+   torch.save(model, "path/to/dir")
+   train.report(checkpoint=Checkpoint.from_directory("path/to/dir"))
+```
+
+Train will make a copy of the specified checkpoint data in the trial dir for persistence. The checkpoint data is managed by Train (e.g., Train may restore only the latest checkpoint or delete previous checkpoints to save space according to checkpoint policy).
+
+The files of the recorded checkpoint can be accessed via `result.checkpoint`. The `Checkpoint` object itself is a logical tuple of `(path, pyarrow.fs.FileSystem)`. Users can also get and set arbitrary metadata to these checkpoints (e.g., preprocessor configs, model settings, etc), which will be recorded in a `metadata.json` file.
```

Comment:
How is the filesystem populated? The user e.g. creates a checkpoint using `Checkpoint.from_directory` - the filesystem is a local filesystem. Ray uploads the checkpoint to the storage path - it becomes a new checkpoint with a S3Filesystem. Can this be customized, e.g. when users need a customized pyarrow-compatible s3 syncer or uses a custom URI?

What happens when the user doesn't set a `storage_path` (so uses a local path) and saves checkpoints? Do we raise an error? As the local filesystem would be invalid on the head node in a distributed setup.

Reply:
The filesystem will always reflect a valid location of the checkpoint data. The Checkpoint class becomes immutable, so if it is uploaded, then a new checkpoint will be created pointing to the remote data. We won't have any to_uri() method, so users shouldn't observe this happening in practice in any case.

Users cannot customize the sync, except by providing an alternative storage_filesystem implementation. We will always use pyarrow.fs.copy_files() to transfer the data.

> What happens when the user doesn't set a storage_path (so uses a local path) and saves checkpoints? Do we raise an error? As the local filesystem would be invalid on the head node in a distributed setup.

Yes, we will raise an error in this case (per FAQ: "If storage path is not set, sync is disabled, and an error is raised if a remote trial tries to persist a checkpoint.").

Reply:
I see, the missing link for me was the `storage_filesystem` setting. This should be sufficient to customize syncing behavior.

It would be good to get a better understanding of which APIs of pyarrow.fs.Filesystem have to be implemented by users to work with Ray AIRs operations, but that's a question for the documentation.